### PR TITLE
Front end

### PR DESCRIPTION
**Pull request by felipe74590**
Started using Streamlit for frontend. So far got to test the events data endpoint to load all event onto the main page. 
Also added some additional clean up that was discussed in other issues.

Commands: ```uv run streamlit run src/main_page.py```

### CODE REVIEW

**Comment by felipe74590** [comment]
image

**Comment by bbelderbos** [comment]
Nice!

File: src/api/main_api.py [hunk 1]
```diff
@@ -50,12 +50,14 @@ def get_saved_events(current_user: User = Depends(get_current_active_user)):
         return events
 
 
-@app.get("/users/me/{user_id}", response_model=User)
+@app.get("/users/me/", response_model=User)
```
Comment (bbelderbos) [inline src/api/main_api.py:54]: I would drop 'me' here because it seems a general user lookup endpoint now

File: src/api/main_api.py [hunk 2]
```diff
@@ -153,9 +155,13 @@ def delete_event(event_id: int):
 @app.post("/users/", response_model=User)
 def create_user(user: User) -> User:
     with Session(engine) as session:
-        session.add(user)
-        session.commit()
-        session.refresh(user)
+        existing_user = session.exec(
+            select(User).where(User.user_name == user.user_name, User.email == user.email)
+        ).first()
+        if existing_user is None:
```
Comment (bbelderbos) [inline src/api/main_api.py]: else show an error?

File: src/constants.py [hunk 3]
```diff
@@ -9,3 +9,11 @@
 DATABASE_URL = config("DATABASE_URL")
 if DATABASE_URL.startswith("postgres://"):
     DATABASE_URL = DATABASE_URL.replace("postgres://", "postgresql://", 1)
+
+SECRET_KEY = config("SECRET_KEY")
+
+EVENTS_API_URL = "https://atx-events-api-7423eb0d8c65.herokuapp.com/"
```
Comment (bbelderbos) [inline src/constants.py]: I would do `config("EVENTS_API_URL", default="https://atx-events-api-7423eb0d8c65.herokuapp.com/")

so you can set it to localhost:8000 for local testing?

File: pyproject.toml [hunk 4]
```diff
@@ -18,6 +19,8 @@ dependencies = [
     "requests>=2.32.3",
     "sqlalchemy>=2.0.34",
     "sqlmodel>=0.0.22",
+    "streamlit>=1.39.0",
+    "tenacity>=9.0.0",
```
Comment (bbelderbos) [inline pyproject.toml:23]: nice find

**Review by bbelderbos — APPROVED** [review]
Nice work

**Comment by felipe74590** [comment]
I added pagination and its still slow, I think its the way I'm calling the data to load onto the page. Edit: pagination is now faster and loads data as page is turned. Need to add page limitation so user can not go to pages past the amount of events provided.

**Comment by bbelderbos** [comment]
I will look at this after lunch, have you looked at the amount of request calls in the server log?

File: src/main_page.py [hunk 5]
```diff
@@ -21,12 +22,27 @@ def load_events(current_count=0, data=None):
...
         current_count += 50
         data.extend(response.json())
-        load_events(current_count, data)
+        get_events(current_count, data)
 
     return data
 
 
-events = load_events()
+def sort_events_by_time(events: list) -> list:
```
Comment (bbelderbos) [inline src/main_page.py]: what about using sorted with generator expression?

```
def sort_events_by_time(events: list) -> list:
    """Sort events by date and return those after the current time."""
    return sorted(
        event for event in events if datetime.fromisoformat(event["start_datetime"]) > datetime.now(),
        key=itemgetter("start_datetime")
    )
```

File: src/main_page.py [hunk 6]
```diff
@@ -21,12 +22,27 @@ def load_events(current_count=0, data=None):
...
+    for index, event in enumerate(events):
+        event_time = datetime.fromisoformat(event["start_datetime"])
+        if event_time > datetime.now():
+            events = events[index:]
+            break
+
+    return events
+
+
+data = get_events()
```
Comment (bbelderbos) [inline src/main_page.py]: add if name == ...

[tip] if __name__ == "__main__"
https://codechalleng.es/tips/if-name-main

**Review by bbelderbos — APPROVED** [review]

File: src/api/main_api.py [hunk 7]
```diff
@@ -172,6 +172,8 @@ def create_user(user: User) -> User:
             session.add(user)
             session.commit()
             session.refresh(user)
+        else:
```
Comment (bbelderbos) [inline src/api/main_api.py]: I would do an early return here:

if user exists:
    raise
# continue with user create logic

File: src/api/main_api.py [hunk 8]
```diff
@@ -75,18 +75,28 @@ def search_events(
     - **to_date**: Filter events till this date
     - **venue_keyword**: Search for events in the venue name.
     """
+    if from_date and to_date and from_date > to_date:
+        raise HTTPException(status_code=400, detail="from_date must be before to_date")
+    elif not from_date and to_date:
```
Comment (bbelderbos) [inline src/api/main_api.py]: can do `if` here, if above already would break out of function

File: src/api/main_api.py [hunk 9]
```diff
@@ -75,18 +75,28 @@ def search_events(
...
+            filters.append(Event.start_datetime <= to_date)
         if venue_keyword is not None:
-            statement = statement.where(Event.venue.ilike(f"%{venue_keyword}%"))
+            filters.append(Event.venue.ilike(f"%{venue_keyword}%"))
         if category_keyword is not None:
-            statement = statement.where(Event.category.ilike(f"%{category_keyword}%"))
+            filters.append(Event.category.ilike(f"%{category_keyword}%"))
+
+        # Apply filters to the statement if there are any
+        if filters:
```
Comment (bbelderbos) [inline src/api/main_api.py:99]: nice

File: src/constants.py [hunk 10]
```diff
@@ -12,7 +12,7 @@
 
 SECRET_KEY = config("SECRET_KEY")
 
-EVENTS_API_URL = "https://atx-events-api-7423eb0d8c65.herokuapp.com/"
+EVENTS_API_URL = config("EVENTS_API_URL", default="https://atx-events-api-7423eb0d8c65.herokuapp.com/")
```
Comment (bbelderbos) [inline src/constants.py]: great

Comment (bbelderbos) [reply src/constants.py]: or maybe default to localhost, to force to setting it in production

**Review by bbelderbos — APPROVED** [review]

**Comment by bbelderbos** [comment]
oops misread, that's great then, indeed it's fast now for me as well:

image

fastapi log:

INFO:     127.0.0.1:51301 - "GET /search_events/?skip=0&limit=50&from_date=2024-10-17+16%3A20%3A07.695016 HTTP/1.1" 200 OK

so all good right?

what's pending? any questions or doubts?

**Comment by felipe74590** [comment]
> what's

Just need to solve the issue of being able to load the next page even though events have run out. For example today events run out on page 11 but I can still go to page 12 and see and empty page, need to fix that bug. After fixing that and issues above I will focus on search bar issue, then user creation and login issues.

**Comment by bbelderbos** [comment]
ok cool and yeah I see, you need to know upfront if you want to make that next call right? 

asked ChatGPT and this seems reasonable to me:

> A common solution to handle pagination in API design is to return a metadata object alongside the paginated data. This metadata can include details such as the total number of items, the current page, the page size (limit), and whether there are more pages available.

What about returning the count then as well?

```
    with Session(engine) as session:
        total_events = session.exec(select(Event)).count()

...

        return {
            "total_count": total_events,
            "events": events
        }
```

File: src/constants.py (hunk 10 above)
Comment (felipe74590) [reply src/constants.py]: I have it vice versa at the moment, but ill flip it.

File: src/api/main_api.py [hunk 11]
```diff
@@ -99,7 +100,10 @@ def search_events(
             statement = statement.where(*filters)
 
         events = session.exec(statement).all()
-        return events
+        count_statment = select(func.count(Event.id)).where(*filters)
+        total_events = session.exec(count_statment).one()
+
+        return {"total_events": total_events, "events": events}
```
Comment (bbelderbos) [inline src/api/main_api.py:106]: cool

File: src/api/main_api.py [hunk 12]
```diff
@@ -168,16 +168,18 @@ def delete_event(event_id: int):
...
 def create_user(user: User) -> User:
+    """Create User in DB, return Error if User already exists in database."""
     with Session(engine) as session:
         existing_user = session.exec(
             select(User).where(User.user_name == user.user_name, User.email == user.email)
         ).first()
-        if existing_user is None:
+        if existing_user:
+            raise HTTPException(status_code=409, detail="User already exists.")
+        else:
```
Comment (bbelderbos) [inline src/api/main_api.py:178]: as there is an early return, you can drop the else and dedent this code block one level

File: src/constants.py [hunk 13]
```diff
@@ -12,7 +12,7 @@
 
 SECRET_KEY = config("SECRET_KEY")
 
-EVENTS_API_URL = config("EVENTS_API_URL", default="https://atx-events-api-7423eb0d8c65.herokuapp.com/")
+EVENTS_API_URL = config("EVENTS_API_URL", default="http://127.0.0.1:8000")
```
Comment (bbelderbos) [inline src/constants.py:15]: yep better

**Pull request merged**